elements/lisk-cryptography/src/encrypt.ts

```diff
@@ -102,38 +107,103 @@ export const decryptMessageWithPassphrase = (
 	crypto.pbkdf2Sync(password, salt, iterations, PBKDF2_KEYLEN, PBKDF2_HASH_FUNCTION);
 
+const getKeyFromPasswordWithArgon2 = async (options: {
+	password: string;
+	salt: Buffer;
+	iterations: number;
+	parallelism: number;
+	memorySize: number;
+}): Promise<Buffer> =>
+	Buffer.from(
+		await argon2id({
+			password: options.password,
+			salt: options.salt,
+			parallelism: options.parallelism,
+			iterations: options.iterations,
+			memorySize: options.memorySize,
+			hashLength: HASH_LENGTH, // we use output size = 32 bytes
+			outputType: 'binary', // we use output binary
+		}),
+	);
+
+export enum Cipher {
+	AES256GCM = 'aes-256-gcm',
+}
+
+export enum KDF {
+	ARGON2 = 'argon2id',
+	PBKDF2 = 'PBKDF2',
+}
+
 export interface EncryptedPassphraseObject {
-	readonly [key: string]: string | number | undefined;
-	readonly cipherText: string;
-	readonly iterations?: number;
-	readonly iv: string;
-	readonly salt: string;
-	readonly tag: string;
 	readonly version: string;
+	readonly ciphertext: string;
+	readonly mac: string;
+	readonly kdf: KDF;
+	readonly kdfparams: {
+		parallelism: number;
+		iterations: number;
+		memorySize: number;
+		salt: string;
+	};
+	readonly cipher: Cipher;
+	readonly cipherparams: {
+		iv: string;
+		tag: string;
+	};
 }
 
-const encryptAES256GCMWithPassword = (
+const encryptAES256GCMWithPassword = async (
 	plainText: string,
 	password: string,
-	iterations: number = PBKDF2_ITERATIONS,
-): EncryptedPassphraseObject => {
+	options?: {
+		kdf?: KDF;
+		kdfparams?: {
+			parallelism?: number;
+			iterations?: number;
+			memorySize?: number;
+		};
+	},
+): Promise<EncryptedPassphraseObject> => {
+	const kdf = options?.kdf ?? KDF.ARGON2;
 	const IV_BUFFER_SIZE = 12;
 	const SALT_BUFFER_SIZE = 16;
-	const iv = crypto.randomBytes(IV_BUFFER_SIZE);
 	const salt = crypto.randomBytes(SALT_BUFFER_SIZE);
-	const key = getKeyFromPassword(password, salt, iterations);
-
+	const iv = crypto.randomBytes(IV_BUFFER_SIZE);
+	const iterations =
+		kdf === KDF.ARGON2 ? ARGON2_ITERATIONS : options?.kdfparams?.iterations ?? PBKDF2_ITERATIONS;
+	const parallelism = options?.kdfparams?.parallelism ?? ARGON2_PARALLELISM;
+	const memorySize = options?.kdfparams?.parallelism ?? ARGON2_MEMORY;
+	const key =
+		kdf === KDF.ARGON2
+			? await getKeyFromPasswordWithArgon2({
+					password,
+					salt,
+					iterations,
+					parallelism,
+					memorySize,
+			  })
+			: getKeyFromPassword(password, salt, iterations);
 	const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
 	const firstBlock = cipher.update(plainText, 'utf8');
 	const encrypted = Buffer.concat([firstBlock, cipher.final()]);
 	const tag = cipher.getAuthTag();
 
 	return {
-		iterations,
-		cipherText: encrypted.toString('hex'),
-		iv: iv.toString('hex'),
-		salt: salt.toString('hex'),
-		tag: tag.toString('hex'),
+		ciphertext: encrypted.toString('hex'),
+		mac: crypto.createHash('sha256').update(key.slice(16, 32)).update(encrypted).digest('hex'),
```

Comment:
better to use function to encapsulate, I think it is already done in @Incede PR

Reply:
I prefer to keep it as is here as it shows clearly what is going on. Also the use-case here is quite specific.